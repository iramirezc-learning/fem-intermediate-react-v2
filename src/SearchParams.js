import React, { useEffect, useState } from "react";
import {
  ANIMALS,
  breeds as fetchBreeds,
  animals as fetchAnimals,
} from "@frontendmasters/pet";
import { connect } from "react-redux";
import Results from "./Results";
import useDropdown from "./useDropdown";
import ThemeDropdown from "./ThemeDropdown";
import { changeTheme, changeLocation } from "./actionCreators";

const SearchParams = ({ theme, setTheme, location, setLocation }) => {
  const [breeds, setBreeds] = useState([]);
  const [animal, AnimalDropdown] = useDropdown("Animal", "dog", ANIMALS);
  const [breed, BreedDropdown, setBreed] = useDropdown("Breed", "", breeds);
  const [pets, setPets] = useState([]);

  useEffect(() => {
    setBreeds([]);
    setBreed("");

    if (!animal) return;

    fetchBreeds(animal)
      .then(({ breeds: breedsFromApi = [] }) => {
        const breedsNames = breedsFromApi.map(({ name }) => name);

        setBreeds(breedsNames);
      })
      .catch(console.error);
  }, [animal, setBreeds, setBreed]);

  async function fetchPets() {
    const { animals: petsFromApi } = await fetchAnimals({
      location,
      breed,
      type: animal,
    });

    setPets(petsFromApi || []);
  }

  return (
    <div className="search-params">
      <form
        onSubmit={(e) => {
          e.preventDefault();
          fetchPets();
        }}
      >
        <ThemeDropdown theme={theme} onChange={setTheme} />
        <label htmlFor="location">
          Location
          <input
            id="location"
            type="text"
            placeholder="Location"
            value={location}
            onChange={(e) => setLocation(e.target.value)}
          />
        </label>
        <AnimalDropdown />
        <BreedDropdown />
        <button type="submit" style={theme.button}>
          Submit
        </button>
      </form>
      <Results pets={pets} />
    </div>
  );
};

const mapStateToProps = ({ theme, location }) => ({
  theme,
  location,
});

const mapDispatchToProps = (dispatch) => ({
  setTheme(theme) {
    dispatch(changeTheme(theme));
  },
  setLocation(location) {
    dispatch(changeLocation(location));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchParams);
