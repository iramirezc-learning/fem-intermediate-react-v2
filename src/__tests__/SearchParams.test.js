import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import PetApi, { ANIMALS, _breeds, _dogs } from "@frontendmasters/pet";
import SearchParams from "../SearchParams";

describe("SearchParams", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = render(<SearchParams />);
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  test("matches the snapshot", () => {
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });

  test("renders the animal dropdown correctly", () => {
    const animalDropdown = wrapper.getByTestId("use-dropdown-animal");

    expect(animalDropdown.children.length).toBe(ANIMALS.length + 1);
  });

  test("renders the breed dropdown correctly", () => {
    const breedDropdown = wrapper.getByTestId("use-dropdown-breed");

    expect(PetApi.breeds).toHaveBeenCalledTimes(1);
    expect(breedDropdown.children.length).toBe(_breeds.length + 1);
  });

  test("renders no pets the first time", () => {
    const searchResults = wrapper.getByTestId("search-results");

    expect(searchResults.textContent).toBe("No pets were found :/");
  });

  test("fetches pets from the PetsApi when Submit is clicked", () => {
    const searchResults = wrapper.getByTestId("search-results");

    fireEvent(wrapper.getByText("Submit"), new MouseEvent("click"));

    expect(PetApi.animals).toHaveBeenCalledTimes(1);
    expect(searchResults.children.length).toBe(_dogs.length);
  });
});
