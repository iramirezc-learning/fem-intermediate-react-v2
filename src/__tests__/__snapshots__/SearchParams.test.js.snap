// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchParams matches the snapshot 1`] = `
<div
  class="search-params"
>
  <form>
    <label
      for="theme"
    >
      Theme
      <select
        id="theme"
      >
        <option
          value="peru"
        >
          Peru
        </option>
        <option
          value="darkblue"
        >
          Dark Blue
        </option>
        <option
          value="chartreuse"
        >
          Chart Reuse
        </option>
        <option
          value="mediumorchid"
        >
          Medium Orchid
        </option>
      </select>
    </label>
    <label
      for="location"
    >
      Location
      <input
        id="location"
        placeholder="Location"
        type="text"
        value="Seattle, WA"
      />
    </label>
    <label
      for="use-dropdown-animal"
    >
      Animal
      <select
        data-testid="use-dropdown-animal"
        id="use-dropdown-animal"
        name="use-dropdown-animal"
      >
        <option
          value=""
        >
          all
        </option>
        <option
          value="dog"
        >
          dog
        </option>
        <option
          value="cat"
        >
          cat
        </option>
        <option
          value="bird"
        >
          bird
        </option>
      </select>
    </label>
    <label
      for="use-dropdown-breed"
    >
      Breed
      <select
        data-testid="use-dropdown-breed"
        id="use-dropdown-breed"
        name="use-dropdown-breed"
      >
        <option
          value=""
        >
          all
        </option>
        <option
          value="Bichon Frise"
        >
          Bichon Frise
        </option>
        <option
          value="Bolognese"
        >
          Bolognese
        </option>
        <option
          value="Bolonka"
        >
          Bolonka
        </option>
        <option
          value="Coton de Tulear"
        >
          Coton de Tulear
        </option>
        <option
          value="Havanese"
        >
          Havanese
        </option>
        <option
          value="Lowchen"
        >
          Lowchen
        </option>
        <option
          value="Maltese"
        >
          Maltese
        </option>
      </select>
    </label>
    <button
      style="background-color: green;"
      type="submit"
    >
      Submit
    </button>
  </form>
  <div
    class="search"
    data-testid="search-results"
  >
    <h1>
      No pets were found :/
    </h1>
  </div>
</div>
`;
